### PR TITLE
implemented protos, client and server

### PR DESCRIPTION
**Pull request by dmitsh**

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		return err
+	}
+	ticker := time.NewTicker(500 * time.Millisecond)
+	offs := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			size := buffer.Len() - offs
```
Comment (smallinsky) [inline cmd/server/main.go]: Race condition. The buffer length is updated by other goroutine during job execution.

File: cmd/server/main.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	if err != nil {
+		return err
+	}
+	ticker := time.NewTicker(500 * time.Millisecond)
+	offs := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
```
Comment (smallinsky) [inline cmd/server/main.go]: Do you see any way to sync output reader without sleeping ?

File: pkg/engine/process.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	m.procMutex.Lock()
+	defer m.procMutex.Unlock()
+	proc, ok := m.procs[uid]
+	if !ok || proc.clientID != clientID {
+		return nil, ErrProcNotFound
+	}
+	return proc.status, nil
+}
+
+func (m *ProcManager) StreamOutputFile(clientID, uid string) (*bytes.Buffer, error) {
```
Comment (smallinsky) [inline pkg/engine/process.go]: `StreamOutputFile` the File prefix suggest that operation will be done based on file where as source byte buffer is used.

File: pkg/engine/process.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	proc.cmd.Stdout = &proc.output
+	proc.cmd.Stderr = &proc.output
+	proc.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	m.addProcess(uid, proc)
+
+	go func() {
+		proc.cmd.Start()
```
Comment (smallinsky) [inline pkg/engine/process.go]: missing error check.

File: pkg/engine/process.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+		m.procMutex.Lock()
+		proc.status.ProcStatus = proto.Status_StatusRunning
+		m.procMutex.Unlock()
+
+		proc.cmd.Wait()
```
Comment (smallinsky) [inline pkg/engine/process.go]: missing error check.

File: cmd/server/main.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+			size := buffer.Len() - offs
+			if size > 0 {
+				err = srv.Send(&proto.LogData{Data: buffer.Bytes()[offs:]})
+				if err != nil {
+					return err
+				}
+				offs += size
+			} else {
+				// no more output - check if the process is still running
+				if status, err := w.procManager.GetProcessStatus(req.GetId()); err != nil || status.ProcStatus != proto.Status_StatusRunning {
```
Comment (smallinsky) [inline cmd/server/main.go]: Race condition on `status.ProcStatus` read. The status returned by is `w.procManager.GetProcessStatus` call is a pointer.

File: cmd/server/main.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	ticker := time.NewTicker(500 * time.Millisecond)
+	offs := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			size := buffer.Len() - offs
+			if size > 0 {
+				err = srv.Send(&proto.LogData{Data: buffer.Bytes()[offs:]})
```
Comment (smallinsky) [inline cmd/server/main.go]: Race condition:  read of shared resource without locking.

File: cmd/server/main.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	ticker := time.NewTicker(500 * time.Millisecond)
+	offs := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			size := buffer.Len() - offs
+			if size > 0 {
```
Comment (smallinsky) [inline cmd/server/main.go]: What if a job produced a new about after `buffer.Len()` call and job status has been changed the `Status_StatusStopped`  ?

File: cmd/server/main.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	log.Println("StopProcess: clientID:", clientID)
+	err := w.procManager.StopProcess(clientID, req.GetId())
+	return &proto.Empty{}, err
+}
+
+func (w *WorkerServer) GetProcessStatus(ctx context.Context, req *proto.JobId) (*proto.Status, error) {
+	clientID := getClientID(ctx)
+	log.Println("GetProcessStatus: clientID:", clientID)
+	status, err := w.procManager.StatusProcess(clientID, req.GetId())
+	return status, err
```
Comment (smallinsky) [inline cmd/server/main.go:95]: race condition the status is a pointer that is modified when job status is updated. So during grpc proto marshaling there will be unprotected read call.

File: cmd/client/main.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	if err != nil {
+		exit(err)
+	}
+
+	fmt.Println("CLN cert:", clientCrtPath, "CLN key:", clientKeyPath, "CA cert:", caKeyPath)
+	fmt.Println(">", cmd, args)
+
+	if err := runClient(cmd, args); err != nil {
+		exit(err)
+		fmt.Printf("error: %v\n", err)
```
Comment (smallinsky) [inline cmd/client/main.go]: fix os.Exit is called during exit(err) this code is unreachable.

File: cmd/server/main.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+		}
+	}
+}
+
+func getClientID(ctx context.Context) string {
+	var clientID string
+	if p, ok := peer.FromContext(ctx); ok {
+		if mtls, ok := p.AuthInfo.(credentials.TLSInfo); ok {
+			for _, item := range mtls.State.PeerCertificates {
+				if txt := item.Subject.String(); strings.HasPrefix(txt, "CN=") {
```
Comment (smallinsky) [inline cmd/server/main.go]: What if for some reason a user was able to issue a cert without CN but with O="`"CN=user1"`"   ?

File: pkg/auth/auth.go [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+	capool := x509.NewCertPool()
+	if !capool.AppendCertsFromPEM(ca) {
+		return nil, fmt.Errorf("failed to add ca certificate")
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{certificate},
+		ClientCAs:    capool,
+	}
```
Comment (smallinsky) [inline pkg/auth/auth.go]: TLS version was not set so the server will allow to connect with old not secure TLS version.

File: pkg/engine/process.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+type ProcManager struct {
+	// process table [process UID : Process]
+	procs     map[string]*Process
+	procMutex sync.Mutex
+
+	// permission table [client ID : permission bitmap]
+	perm map[string]int
+
+	// process UID
+	uid string
```
Comment (smallinsky) [inline pkg/engine/process.go]: Where this filed is used ?

File: pkg/engine/process.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+		m.procMutex.Lock()
+		proc.status.ProcStatus = proto.Status_StatusRunning
+		m.procMutex.Unlock()
+
+		proc.cmd.Wait()
+
+		m.procMutex.Lock()
+		proc.status.ProcStatus = proto.Status_StatusStopped
+
+		osStatus := proc.cmd.ProcessState.Sys().(syscall.WaitStatus)
```
Comment (smallinsky) [inline pkg/engine/process.go]: missing case check result.

File: cmd/runner/main.go [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+	cmd := exec.Command("/proc/self/exe", append([]string{"cgr"}, os.Args[2:]...)...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWPID,
+	}
+
+	cmd.Run()
```
Comment (smallinsky) [inline cmd/runner/main.go]: missing error check

File: cmd/runner/main.go [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+		err = errors.New("invalid input")
+	}
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func start() error {
+	cmd := exec.Command("/proc/self/exe", append([]string{"cgr"}, os.Args[2:]...)...)
+	cmd.Stdin = os.Stdin
```
Comment (smallinsky) [inline cmd/runner/main.go]: Thus the user command input is not supported the stdin chan can be skipped.

File: cmd/runner/main.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	// set cgroup procs id
+	if err := writeInt(cgroupMemProcsFile, os.Getpid()); err != nil {
+		return err
+	}
+	if err := writeInt(cgroupCpuProcsFile, os.Getpid()); err != nil {
+		return err
+	}
+
+	cmd := exec.Command(os.Args[3], os.Args[4:]...)
```
Comment (smallinsky) [inline cmd/runner/main.go:84]: Referencing to Args[3] will panic when a user command was empty.

File: cmd/runner/main.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	cmd.Run()
```
Comment (smallinsky) [inline cmd/runner/main.go]: missing error check.

File: cmd/runner/main.go [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+	os.Exit(cmd.ProcessState.Sys().(syscall.WaitStatus).ExitStatus())
```
Comment (smallinsky) [inline cmd/runner/main.go]: Missing cast result check. This ProcessState can be empty if cmd.Start call failed.

File: cmd/runner/main.go [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+		return err
+	}
+	if err := os.MkdirAll(cgroupCpuDir, 0755); err != nil {
+		return err
+	}
+	// set memory limit
+	if err := writeInt(cgroupMemLimitFile, rssLimit*1024*1024); err != nil {
+		return err
+	}
+	if err := writeInt(cgroupMemSwapLimitFile, 0); err != nil {
```
Comment (smallinsky) [inline cmd/runner/main.go]: A bit Inconsistency: In this call  the `0` number is used as PID  where if next calls the explicit process id obtained by os.GetPid  is used.

**Review by smallinsky — COMMENTED** [review]
I have leaf some comments. Also I don't see any tests.

File: cmd/runner/main.go (hunk 20 above)
Comment (dmitsh) [reply cmd/runner/main.go]: oh, right, I was using it for testing and forgot to remove. Removing now.

File: cmd/runner/main.go (hunk 17 above)
Comment (dmitsh) [reply cmd/runner/main.go:84]: Added check for empty executable path in WorkerServer.StartProcess

File: cmd/server/main.go (hunk 2 above)
Comment (dmitsh) [reply cmd/server/main.go]: yes, I was trying to do it the easy way :)
fixed

File: cmd/server/main.go (hunk 6 above)
Comment (dmitsh) [reply cmd/server/main.go]: I refactored this code

File: cmd/server/main.go (hunk 11 above)
Comment (dmitsh) [reply cmd/server/main.go]: to be honest, I don't know :)

File: pkg/engine/rwbuffer.go [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+func (r *BufReader) Read(p []byte) (int, error) {
+	r.writer.cond.L.Lock()
+	defer r.writer.cond.L.Unlock()
+
+	if r.offs < len(r.writer.data) {
+		size := min(len(r.writer.data)-r.offs, len(p))
+		newoffs := r.offs + size
+		copy(p, r.writer.data[r.offs:newoffs])
+		r.offs = newoffs
+		return size, nil
```
Comment (smallinsky) [inline pkg/engine/rwbuffer.go]: the `copy` function already leverages the min(len(dst), len(src)): 
https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/builtin/builtin.go;l=139

so I think that this code can be simplified to: 
```suggestion
		
		n := copy(p, r.writer.data[r.offs:])
		r.off += n
		return n

```

File: cmd/server/main.go (hunk 11 above)
Comment (smallinsky) [reply cmd/server/main.go]: The logic should not check CN based  on combined Subject String result because this will allow for potential abuse this and encode `CN=user1` value in other DN fields like Organisation.

File: cmd/client/main.go [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+)
+
+var (
+	clientCrtPath, clientKeyPath, caKeyPath string
+)
+
+func init() {
+	flag.StringVar(&clientCrtPath, "cln.crt", "", "client certificate filepath")
+	flag.StringVar(&clientKeyPath, "cln.key", "", "client key filepath")
+	flag.StringVar(&caKeyPath, "ca.crt", "", "CA certificate path")
```
Comment (zmb3) [inline cmd/client/main.go]: `caKeyPath` is a strange name for a CA _certificate_

File: cmd/client/main.go [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+		}
+		for {
+			resp, err := stream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			fmt.Println(string(resp.GetData()))
```
Comment (zmb3) [inline cmd/client/main.go]: You don't really need the string conversion, nor the functionality of the `fmt` package.

`os.Stdout.Write(resp.GetData())` would be more efficient.

File: cmd/runner/main.go [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+func main() {
+	var err error
+
+	switch os.Args[1] {
+	case "start":
+		err = start()
+	case "cgr":
+		err = cgr()
+	default:
+		err = errors.New("invalid input")
```
Comment (zmb3) [inline cmd/runner/main.go]:
```suggestion
		err = fmt.Errorf("invalid command %q", os.Args[1])
```

File: cmd/runner/main.go [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+// (2) including block device numbers (currently omitted) and setting associated bandwidth limits
+const (
+	cpuShares = 512 // CPU shares
+	rssLimit  = 10  // memory limit with MB
+)
+
+func main() {
+	var err error
+
+	switch os.Args[1] {
```
Comment (zmb3) [inline cmd/runner/main.go:24]: Should probably check that `len(os.Args) > 1` before accessing element 1.

File: cmd/runner/main.go [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+	// set CPU shares
+	if err := writeInt(cgroupCpuSharesFile, cpuShares); err != nil {
+		return err
+	}
+	// set cgroup procs id
+	if err := writeInt(cgroupMemProcsFile, os.Getpid()); err != nil {
+		return err
+	}
+	if err := writeInt(cgroupCpuProcsFile, os.Getpid()); err != nil {
+		return err
```
Comment (zmb3) [inline cmd/runner/main.go:81]: In all of these error cases, consider cleaning up the partially-completed cgroup configuration before returning.

File: cmd/runner/main.go [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	check(cmd.Run())
+
+	return nil
+}
+
+func writeInt(path string, value int) error {
+	return ioutil.WriteFile(path, []byte(fmt.Sprintf("%d", value)), 0755)
```
Comment (zmb3) [inline cmd/runner/main.go]: The ioutil package is deprecated, and `Itoa` is more efficient than `Sprintf` for a simple conversion like this.

```suggestion
	return os.WriteFile(path, []byte(strconv.Itoa(value)), 0755)
```

File: cmd/server/main.go [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+			return nil
+		default:
+			n, err := reader.Read(data)
+			if err != nil {
+				if err == io.EOF {
+					return nil
+				}
+				return err
+			}
+			if n > 0 {
```
Comment (zmb3) [inline cmd/server/main.go:120]: You may want to check if `n > 0` and send any data before checking the error. I haven't got to your streamer implementation yet, but if it conforms to `io.Reader` then it's totally possible for `n > 0 && err != nil` to be true.

File: pkg/auth/auth.go [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+	"google.golang.org/grpc/credentials"
+)
+
+func GetTLS(crt, key, caCrt string, isServer bool) (credentials.TransportCredentials, error) {
+	certificate, err := tls.LoadX509KeyPair(crt, key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load certificate: %v", err)
+	}
+
+	ca, err := ioutil.ReadFile(caCrt)
```
Comment (zmb3) [inline pkg/auth/auth.go]:
```suggestion
	ca, err := os.ReadFile(caCrt)
```

File: pkg/auth/auth.go [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to load CA certificate: %v", err)
+	}
+
+	capool := x509.NewCertPool()
+	if !capool.AppendCertsFromPEM(ca) {
+		return nil, fmt.Errorf("failed to add ca certificate")
```
Comment (zmb3) [inline pkg/auth/auth.go]:
```suggestion
		return nil, errors.New("failed to add ca certificate")
```

File: pkg/auth/auth.go [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{certificate},
+		ClientCAs:    capool,
+
+		MinVersion: tls.VersionTLS12,
```
Comment (zmb3) [inline pkg/auth/auth.go:33]: Why TLS 1.2?

File: pkg/engine/process.go [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+	if err := m.checkPermission(clientID, PermStatus); err != nil {
+		return nil, err
+	}
+	m.procMutex.Lock()
+	defer m.procMutex.Unlock()
+	proc, ok := m.procs[uid]
+	if !ok || proc.clientID != clientID {
+		return nil, ErrProcNotFound
+	}
+	return proc.status, nil
```
Comment (zmb3) [inline pkg/engine/process.go]: I would not return `proc.status` here as you can't make any guarantees that the caller will access it safely.

File: pkg/engine/rwbuffer.go [hunk 33]
```diff
@@ -0,0 +1,80 @@
...
+	writer.cond = sync.NewCond(writer)
+	return writer
+}
+
+func (w *BufWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+	data := make([]byte, len(p))
+	copy(data, p)
```
Comment (zmb3) [inline pkg/engine/rwbuffer.go]: Why is this necessary? Can't you copy directly from `p` into `w.data`?

File: pkg/engine/rwbuffer.go [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+func NewBufWriter() *BufWriter {
+	writer := &BufWriter{}
+	writer.cond = sync.NewCond(writer)
+	return writer
+}
+
+func (w *BufWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
```
Comment (zmb3) [inline pkg/engine/rwbuffer.go:36]: Consider also making it an error to write to `w` if `w.closed` is true.

File: pkg/engine/rwbuffer.go [hunk 35]
```diff
@@ -0,0 +1,80 @@
...
+		return size, nil
+	}
+
+	if r.writer.closed {
+		return 0, io.EOF
+	}
+
+	r.writer.cond.Wait()
+
+	return 0, nil
```
Comment (zmb3) [inline pkg/engine/rwbuffer.go:73]: `0, nil` is okay, as it should force the caller to call `Read` again.

However, it's likely that you have new data that you could return now rather than reading 0 bytes and forcing the caller to retry.

File: cmd/client/main.go (hunk 23 above)
Comment (zmb3) [reply cmd/client/main.go]: Additionally, `Println` is going to add a `\n`, so the output the user sees on their console is not going to match the actual output of the process.

File: pkg/engine/rwbuffer.go (hunk 35 above)
Comment (dmitsh) [reply pkg/engine/rwbuffer.go:73]: @zmb3, I think Go does not support reentrance under lock. If I call `Read` recursively, it will cause deadlock, isn't it?

Comment (zmb3) [reply pkg/engine/rwbuffer.go:73]: Yes, that would deadlock.

But there's no reason why you could not loop or have an internal read call that assumes the lock is already held.

Comment (dmitsh) [reply pkg/engine/rwbuffer.go:73]: agree!

File: cmd/server/main.go (hunk 28 above)
Comment (dmitsh) [reply cmd/server/main.go:120]: Is the rationale here to send the data quicker?

File: cmd/server/main.go (hunk 11 above)
Comment (dmitsh) [reply cmd/server/main.go]: got it, thanks!
